### PR TITLE
Move safe_auth_ffi to a separate crate - Rebased

### PR DESCRIPTION
**Pull request by ravinderjangra**

### CODE REVIEW

**Comment by ravinderjangra** [comment]
Updating scripts to include `safe_authenticator_ffi` checks, builds and tests.

File: safe_authenticator/src/ipc.rs [hunk 1]
```diff
@@ -81,8 +79,11 @@ pub fn decode_ipc_msg(
                         })),
                         AppState::Revoked | AppState::NotAuthenticated => {
                             // App is not authenticated
-                            let (error_code, description) =
-                                ffi_error!(FFIError::from(IpcError::UnknownApp));
+                            // let (error_code, description) =
```
Comment (lionel1704) [inline safe_authenticator/src/ipc.rs]: This can be removed (minor point)

File: safe_authenticator_ffi/src/ffi/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use crate::ffi::apps::AppPermissions;
+use ffi_utils::{vec_into_raw_parts, ReprC};
+use safe_authenticator::apps::RegisteredApp as NativeRegisteredApp;
+use safe_core::ipc::req::{containers_from_repr_c, containers_into_vec};
+use safe_core::ipc::AppExchangeInfo as NativeAppExchangeInfo;
+use safe_core::ipc::IpcError;
+use safe_nd::AppPermissions as NativeAppPermissions;
+
+/// Registered native app converter.
+#[allow(unsafe_code)]
```
Comment (lionel1704) [inline safe_authenticator_ffi/src/ffi/helpers.rs]: Can be done at crate level ( minor point )

File: safe_authenticator_ffi/src/test_utils.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Provides utilities to test the authenticator functionality.
+
+    #![allow(clippy::not_unsafe_ptr_arg_deref, unsafe_code)]
```
Comment (lionel1704) [inline safe_authenticator_ffi/src/test_utils.rs]: Unnecessary tab (minor)

File: safe_core/src/errors.rs [hunk 4]
```diff
@@ -172,6 +174,94 @@ impl Debug for CoreError {
     }
 }
 
+/// TODO
```
Comment (lionel1704) [inline safe_core/src/errors.rs]: Write function doc

File: safe_core/src/errors.rs [hunk 5]
```diff
@@ -172,6 +174,94 @@ impl Debug for CoreError {
...
+        CoreError::OperationAborted => ERR_OPERATION_ABORTED,
+        CoreError::SelfEncryption(_) => ERR_SELF_ENCRYPTION,
+        CoreError::RequestTimeout => ERR_REQUEST_TIMEOUT,
+        CoreError::ConfigError(_) => ERR_CONFIG_FILE,
+        CoreError::IoError(_) => ERR_IO,
+        CoreError::Unexpected(_) => ERR_UNEXPECTED,
+    }
+}
+
+/// TODO
```
Comment (lionel1704) [inline safe_core/src/errors.rs]: Write doc comment

**Pull request merged**